firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +303,250 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
+        # Works by tabulating the coarse 1D Lagrange basis functions in the (Nf+1)-by-(Nc+1) matrix Jhat,
+        # and using the fact that the 2d/3d tabulation is the tensor product J = kron(Jhat, kron(Jhat, Jhat))
+        ndim = Vf.ufl_domain().topological_dimension()
+        zf = self.get_nodes_1d(Vf)
+        zc = self.get_nodes_1d(Vc)
+        Jnp = self.barycentric(zc, zf)
+        # Declare array shapes to be used as literals inside the kernels, I follow to the m-by-n convention.
+        (mx, nx) = Jnp.shape
+        (my, ny) = (mx, nx)
+        (mz, nz) = (mx, nx) if ndim == 3 else (1, 1)
+        nscal = np.prod(Vf.shape, dtype=int)  # number of components
+        mxyz = mx*my*mz  # dim of Vf scalar element
+        nxyz = nx*ny*nz  # dim of Vc scalar element
+        lwork = nscal*max(mx, nx)*max(my, ny)*max(mz, nz)  # size for work arrays
+
+        Jhat = np.ascontiguousarray(Jnp.T)  # BLAS uses FORTRAN ordering
+        # Pass the 1D tabulation as hexadecimal string
+        JX = ', '.join(map(float.hex, np.matrix.flatten(Jhat)))
+        # The Kronecker product routines assume 3D shapes, so in 2D we pass one instead of Jhat
+        JZ = "JX" if ndim == 3 else "&one"
+
+        # Common kernel to compute y = kron(A3, kron(A2, A1)) * x
+        # Vector and tensor field genearalization from Deville, Fischer, and Mund section 8.3.1.
+        kronmxv_code = """
+        extern void dgemm_(char *TRANSA, char *TRANSB, int *m, int *n, int *k,
+              double *alpha, double *A, int *lda, double *B, int *ldb,
+              double *beta , double *C, int *ldc);
+
+        void kronmxv(int tflag,
+            int mx, int my, int mz,
+            int nx, int ny, int nz, int nel,
+            double *A1, double *A2, double *A3,
+            double *x , double *y){
+
+        /*
+        Kronecker matrix-vector product
+
+        y = op(A) * x,  A = kron(A3, kron(A2, A1))
+
+        where:
+        op(A) = transpose(A) if tflag>0 else A
+        op(A1) is mx-by-nx,
+        op(A2) is my-by-ny,
+        op(A3) is mz-by-nz,
+        x is (nx*ny*nz)-by-nel,
+        y is (mx*my*mz)-by-nel.
+
+        Important notes:
+        The input data in x is destroyed in the process.
+        Need to allocate nel*max(mx, nx)*max(my, ny)*max(mz, nz) memory for both x and y.
+        */
+
+        int m,n,k,s,p,lda;
+        char TA1, TA2, TA3;
+        char tran='T', notr='N';
+        double zero=0.0E0, one=1.0E0;
+
+        if(tflag>0){
+           TA1 = tran;
+           TA2 = notr;
+        }else{
+           TA1 = notr;
+           TA2 = tran;
+        }
+        TA3 = TA2;
+
+        m = mx;  k = nx;  n = ny*nz*nel;
+        lda = (tflag>0)? nx : mx;
+        dgemm_(&TA1, &notr, &m,&n,&k, &one, A1,&lda, x,&k, &zero, y,&m);
+
+        p = 0;  s = 0;
+        m = mx;  k = ny;  n = my;
+        lda = (tflag>0)? ny : my;
+        for(int i=0; i<nz*nel; i++){
+           dgemm_(&notr, &TA2, &m,&n,&k, &one, y+p,&m, A2,&lda, &zero, x+s,&m);
```

Comment:
It's more work, but I'd kind of be interested to know what the performance looks like if you use [libxsmm](https://libxsmm.readthedocs.io/en/latest/libxsmm_mm/#overview) for these small dgemms. I think it should be right in the ballpark where one would expect a decent improvement.